62_CAD/DrawResourcesFiller.cpp

```diff
@@ -282,194 +453,758 @@ void DrawResourcesFiller::drawFontGlyph(
+					// It's crucial to deallocate this index to avoid leaks and preserve descriptor pool space.
+					// No semaphore wait needed here, as the GPU never got to use this slot.
+					suballocatedDescriptorSet->multi_deallocate(imagesArrayBinding, 1u, &cachedImageRecord->arrayIndex, {});
+					cachedImageRecord->arrayIndex = InvalidTextureIndex;
+				}
+				
+				// erase the entry we failed to fill, no need for `evictImage_SubmitIfNeeded`, because it didn't get to be used in any submit to defer it's memory and index deallocation
+				imagesCache->erase(imageID);
+			}
+		}
 		else
 		{
-			// TODO: Log
-			success = false;
+			m_logger.log("ensureGeoreferencedImageAvailability_AllocateIfNeeded failed index allocation. shouldn't have happened.", nbl::system::ILogger::ELL_ERROR);
+			cachedImageRecord->arrayIndex = InvalidTextureIndex;
 		}
 	}
-	return success;
+
+	
+	// cached or just inserted, we update the lastUsedFrameIndex
+	cachedImageRecord->lastUsedFrameIndex = currentFrameIndex;
+
+	assert(cachedImageRecord->arrayIndex != InvalidTextureIndex); // shouldn't happen, because we're using LRU cache, so worst case eviction will happen + multi-deallocate and next next multi_allocate should definitely succeed
+	return (cachedImageRecord->arrayIndex != InvalidTextureIndex);
 }
 
-bool DrawResourcesFiller::finalizeLineStyleCopiesToGPU(SIntendedSubmitInfo& intendedNextSubmit)
+bool DrawResourcesFiller::queueGeoreferencedImageCopy_Internal(image_id imageID, const StreamedImageCopy& imageCopy)
 {
-	bool success = true;
-	// Copy LineStyles
-	uint32_t remainingLineStyles = currentLineStylesCount - inMemLineStylesCount;
-	SBufferRange<IGPUBuffer> stylesRange = { sizeof(LineStyle) * inMemLineStylesCount, sizeof(LineStyle) * remainingLineStyles, gpuDrawBuffers.lineStylesBuffer };
-	if (stylesRange.size > 0u)
-	{
-		const LineStyle* srcLineStylesData = reinterpret_cast<LineStyle*>(cpuDrawBuffers.lineStylesBuffer->getPointer()) + inMemLineStylesCount;
-		if (m_utilities->updateBufferRangeViaStagingBuffer(intendedNextSubmit, stylesRange, srcLineStylesData))
-			inMemLineStylesCount = currentLineStylesCount;
-		else
+	auto& vec = streamedImageCopies[imageID];
+	vec.emplace_back(imageCopy);
+	return true;
+}
+
+// TODO[Przemek]: similar to other drawXXX and drawXXX_internal functions that create mainobjects, drawObjects and push additional info in geometry buffer, input to function would be a GridDTMInfo
+// We don't have an allocator or memory management for texture updates yet, see how `_test_addImageObject` is being temporarily used (Descriptor updates and pipeline barriers) to upload an image into gpu and update a descriptor slot (it will become more sophisticated but doesn't block you)
+void DrawResourcesFiller::drawGridDTM(
+	const float64_t2& topLeft,
+	float64_t2 worldSpaceExtents,
+	float gridCellWidth,
+	uint64_t textureID,
+	const DTMSettingsInfo& dtmSettingsInfo,
+	SIntendedSubmitInfo& intendedNextSubmit,
+	bool drawGridOnly/* = false*/)
```

Comment:
remove this param and use dtmSettings mode instead. if it's OUTLINE then it's GRID only.
and texture ID isn't even considered. make sure that's documented in DrawResources.h.